### DISCUSSION TITLE
Shared interface between CapsuleContainer and CapsuleReader

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#219] [General]
Do you think would be possible (and a good idea) to have a common interface between `CapsuleContainer` and `CapsuleReader`?

My use case is very simple. I have to duplicate every entry on my extensions on both types:

```dart
extension CapsuleContainerExt on CapsuleContainer {
  Language get language => read(languageCapsule);
}

extension CapsuleReaderExt on CapsuleReader {
  Language get language => call(languageCapsule);
}
```

**Comment by GregoryConrad** [comment]
You shouldn't really be reading stuff directly from the `CapsuleContainer` unless you're writing a Dart-only app and/or are doing some weird edge case stuff where you can't use a regular capsule (routing in Flutter comes to mind with some imperative APIs).

And seeing as `CapsuleReader` interface is set in stone from v1 release, and only exposes the `call()` function, it wouldn't make sense to write something like:

```dart
final container = CapsuleContainer();
final something = container(someCapsule);
```

Besides, not sure I'd really recommend adding extensions onto `CapsuleReader` anyways--too easy to get them confused with side effects since that will be the same syntax. I.e.,

```dart
final lang = use.language;
```

Looks like it is a side effect, possibly a `WidgetSideEffect` that uses `BuildContext` to access a localization preference. You shouldn't mix capsule reads in there

**Reply by busslina** [reply]
I understand.

> You shouldn't really be reading stuff directly from the CapsuleContainer unless you're writing a Dart-only app...

My use-case is backend side, but also flutter side when using external libraries which are not prepared to be integrated into ReArch declarative flow. E.G. Beamer Router.

> easy to get them confused with side effects since that will be the same syntax

This could be solved by using some identifying syntax like putting and end character (`use.language_`).
Also, I only use this technique with a few capsules that are used a lot, making them easier to type.

I fully understand you point, but I don't get to the point of 100% declarative, even if there is no external library forcing me in the opposite direction. There is always cases where I have to use the container... But of course I will continue to improve my coding willing to be the closest to that 100%.

:)

**Comment by busslina** [comment]
> too easy to get them confused with side effects since that will be the same syntax

Yes, but in my POV that is a limited way to see it. We can do mistakes everywhere, but it really helps in readability and fast coding once you get used to (above all on capsules that are read many times). Still annoying me to write two mirror extensions... Is still impossible to have only one common extension?

![imagen](https://github.com/user-attachments/assets/b112a9c8-e6fd-4fbf-8a1e-df91e25b97ed)

**Reply by GregoryConrad** [reply]
I don’t see why you’re reading out of the container—that should be exceptionally rare. 

You’re welcome to make extensions in your own code, but be warned anyone else reading it might get confused since it’s not idiomatic (use.something is reserved for side effects everywhere else in the implementation)

**Reply by GregoryConrad** [reply]
And also, when macros come out, you can always make a macro you apply on capsules that generates both extensions for you

**Reply by busslina** [reply]
> I don’t see why you’re reading out of the container—that should be exceptionally rare.

Backend apps (mainly servers or sysadmin apps).

**Reply by busslina** [reply]
Maybe I'm still fighting against declarative on my backend apps. I'm gonna check it...
Seems a battle between declarative and inheritance/composition (I suppose I have to learn a lot yet).
I just started to do "my 95% inside capsule" methodology hehe

**Reply by GregoryConrad** [reply]
> Backend apps (mainly servers or sysadmin apps).

even then, essentially all of your logic should be within capsules. The only thing that shouldn’t be within capsules is “glue” code that links up the capsule with the libraries you’re using

**Comment by busslina** [comment]
Do you think this is a legit case to use `CapsuleContainer`?

An http server has an entry handler. For now, my entry handler is not a capsule and it's reading from `CapsuleContainer` each time it's invoked (each incoming request), so it's obtaining the updated subhandler capsules.

But, what if I try to make it a capsule? I think the http `serve` method caches the first version of the entry handler capsule and will not use its latest updated version on the future.

```dart
RestApiServerController restApiServerCapsule(CapsuleHandle use) {
  final config = use(configCapsule);

  Future<void> start() async {
    final restApiConfig = config.asServerConfig.restApi!;

    final localPort = restApiConfig.localListenPort;
    final publicHost = restApiConfig.publicListenHost;
    final publicPort = restApiConfig.publicListenPort;

    final server = await serve(
      restApiEntryHandler,
      InternetAddress.anyIPv4,
      localPort,
    );

    print('Serving at http://${server.address.host}:${server.port}');
    print(
      'Serving at https://$publicHost${publicPort == 443 ? '' : ':$publicPort'}',
    );
  }

  return RestApiServerController(
    start: start,
  );
}

FutureOr<Response> restApiEntryHandler(Request request) async {
  // Websocket
  {
    final res = await websocketEntryHandler(request);
    if (res != null) {
      return res;
    }
  }

  // Entry router
  return cc.read(_routerCapsule).call(request);
}

Router _routerCapsule(CapsuleHandle use) =>
    Router()..mount('/api/v1', use(apiV1RouterCapsule).call);
```

**Reply by GregoryConrad** [reply]
> Do you think this is a legit case to use `CapsuleContainer`?

No; I'd use the action pattern: https://rearch.gsconrad.com/paradigms/actions

```dart
FutureOr<Response> Function(Request) restApiEntryHandlerCapsule(CapsuleHandle use) {
  final route = use(_routerCapsule);
  return (request) {
    return await websocketEntryHandler(request) ?? route(request);
  };
}
```

**Reply by busslina** [reply]
What I mean, is that the http `serve` method is called once, so even if we put the top level entry on a capsule, it will be read only its value at the moment of the `serve` being called.

Gonna check if I can find a solution based on the action pattern.

Updated:
You can see here (on the left) that `restApiEntryHandler` can be updated but will be ignored as `start`/`serve` is executed once. It will be using the cached `router` (on the right), not the latest version:
![imagen](https://github.com/user-attachments/assets/4bf38727-85a1-428d-aca9-6035247ef1ac)

**Reply by busslina** [reply]
Just for the case this last comment got lost/zombie :)

**Reply by GregoryConrad** [reply]
Is there a question? Sorry, bit lost. I'd keep things in capsules like that though.

**Reply by busslina** [reply]
I show you an scenario where seems to be needed to access via the `CapsuleContainer`in order to get the latest version. Th question would be: Am I right or maybe I'm missing something?

I'm creating an discussion/issue about dynamic capsules...

**Reply by GregoryConrad** [reply]
> I show you an scenario where seems to be needed to access via the CapsuleContainer in order to get the latest version.

If you need to get the latest value, make a new capsule and `use` the appropriate capsules at the top of the new one, and that way it will always be rebuilt when the dependencies change.

If for some reason you can't do that, use container listeners to interact with any imperative APIs to update state. The only place you should be using a `CapsuleContainer` is in main, to setup your application.

**Reply by busslina** [reply]
I just did a video trying to clarify this, as seems I'm not expressing it well.

https://streamable.com/ayvkv3

**Reply by GregoryConrad** [reply]
Thanks, I think I see what you're saying. The problem comes if `configCapsule` is updated and you need an updated copy of that in the handler?

Without thinking too in depth about it, here's my initial thought:

```dart
Future<void> main() async {
  final container = CapsuleContainer();
  // ...
  await serve(
      (request) => container.read(restApiHandlerCapsule)(request),
      InternetAddress.anyIPv4,
      localPort,
    );
}
```

**Reply by GregoryConrad** [reply]
(the above will use the latest restApiHandler whenever a new request is made, while also keeping the scope of the CapsuleContainer down)
